src/services/job-processor/service.ts

```diff
@@ -68,56 +73,60 @@ export const makeJobProcessor = ({
-        fromBlock,
-        toBlock,
-        operatorId
-      )
-
-      logger.info('Handling ejection requests', {
-        amount: eventsForEject.length,
-      })
-
-      for (const [ix, event] of eventsForEject.entries()) {
-        logger.debug(`Handling exit ${ix + 1}/${eventsForEject.length}`, event)
-
-        try {
-          if (await consensusApi.isExiting(event.validatorPubkey)) {
-            logger.info('Validator is already exiting(ed), skipping')
-            continue
-          }
+    const lastBlockNumber = await executionApi.latestBlockNumber()
 
-          if (config.DRY_RUN) {
-            logger.info('Not initiating an exit in dry run mode')
-            continue
-          }
+    const eventsForEject = await exitLogs.getLogs(operatorIds, lastBlockNumber)
 
-          if (config.VALIDATOR_EXIT_WEBHOOK) {
-            await webhookProcessor.send(config.VALIDATOR_EXIT_WEBHOOK, event)
-          } else {
-            await messagesProcessor.exit(messageStorage, event)
+    logger.info('Handling ejection requests', {
+      amount: eventsForEject.length,
+    })
+
+    for (const [ix, event] of eventsForEject.entries()) {
+      logger.debug(`Handling exit ${ix + 1}/${eventsForEject.length}`, event)
+
+      if (event.acknowledged) {
+        logger.debug('Event already acknowledged, skipping')
+        continue
+      }
+
+      try {
+        if (await consensusApi.isExiting(event.validatorPubkey)) {
+          logger.info('Validator is already exiting(ed), skipping', {
+            validatorIndex: event.validatorIndex,
+          })
+          // Acknowledge the event to avoid processing it again
+          // We do an additional check, because if we didn't check for finalized,
+          // we might miss the reorganization.
+          if (
+            lastBlockNumber - event.blockNumber > FINALIZED_BLOCK_EQUIVALENT ||
+            (await consensusApi.isExiting(event.validatorPubkey, 'finalized'))
+          ) {
+            event.ack()
           }
-        } catch (e) {
-          logger.error(`Unable to process exit for ${event.validatorPubkey}`, e)
-          metrics.exitActions.inc({ result: 'error' })
+          continue
+        }
+
+        if (config.DRY_RUN) {
+          logger.info('Not initiating an exit in dry run mode')
+          continue
+        }
+
+        if (config.VALIDATOR_EXIT_WEBHOOK) {
+          await webhookProcessor.send(config.VALIDATOR_EXIT_WEBHOOK, event)
+        } else {
+          await messagesProcessor.exit(messageStorage, event)
         }
+      } catch (e) {
+        logger.error(`Unable to process exit for ${event.validatorPubkey}`, e)
+        metrics.exitActions.inc({ result: 'error' })
       }
 
       logger.info('Updating exit messages left metrics from contract state')
       try {
         const lastRequestedValIx =
-          await executionApi.lastRequestedValidatorIndex(operatorId)
+          await exitLogs.verifier.lastRequestedValidatorIndex(
```

Comment:
FYI: This method will be removed in TE upgrade

Reply:
https://linear.app/lidofi/issue/ORC-355/validatorejector-tw-preparations